### DISCUSSION TITLE
In manhattan routing, link path overlaps whenever routes are same.

### DISCUSSION THREAD

**Discussion by csruchitaoza** [clientIO/joint#2711] [Q&A] [answered]
### Introduction

I need the following functionality:
1. Smart routing algorithm that avoids other cells while drawing a line
2. The line should not overlap with other line

Tried solutions :
Multiple connection siblings ( but I still need smart routing )
ELK ( In this algorithm also there are some colids)
Jump overs ( after adding a jump, the line is going to the same path )
Other routings ( manhattan seems a good option but it overlaps the line)

A quick answer would be helpful. Thank you.

Using tech - Angular 17, JointJs/Core

![image](https://github.com/user-attachments/assets/4cd65227-8b51-409c-b5a3-4fc71c810f65)

### Does your question relate to JointJS or JointJS+. Select both if applicable.

JointJS

**Comment by kumilingus** [comment, marked as answer]
The `manhattan` router does not avoid other links.

You can check `libavoid` library example, [here](https://www.jointjs.com/demos/libavoid-standalone-link-routing).

**Reply by csruchitaoza** [reply]
Is the demo available in TypeScript? I tried to covert it to typeScript with the following steps

1. Installed libavoid-js and
2. Added avoid-router.ts with conversion to TS
3. In Component, configured avoid-router like in the demo
4. In angular.json, I added this under assets

`
{
    "input": "node_modules/libavoid-js/dist",
    "glob": "libavoid.wasm",
    "output": ""
}
`

However, the router is not working. Am I missing some point?

Gives error :
: Aborted(both async and sync fetching of the wasm failed)
GET http://localhost:4200/libavoid.wasm 404 (Not Found)

If there's a demo in angular or react that would be helpful.

**Reply by kumilingus** [reply]
We only have a JavaScript version of this demo.

It seems like this:

`{ "input": "node_modules/libavoid-js/dist", "glob": "libavoid.wasm", "output": "" }`

not doing its job:

`GET http://localhost:4200/libavoid.wasm 404 (Not Found)`

You can try to copy the `libavoid.wasm` to the root of the localhost manually and see if it works.

**Reply by csruchitaoza** [reply]
Tried that also but it is giving same error

**Reply by kumilingus** [reply]
You copied the file to the root directory of the localhost server and it still complains that the file was not found when requested via `GET http://localhost:4200/libavoid.wasm`?

If you open the `http://localhost:4200/libavoid.wasm` in the browser what does it say?

**Reply by csruchitaoza** [reply]
Cannot GET /libavoid.wasm

Tried in Src folder, in assert folder and outside of src folder

**Reply by kumilingus** [reply]
I think it's the `dist` folder the files are served from in _Angular_.

**Reply by csruchitaoza** [reply]
Hi. I solved that error by providing a path in the Load function

But there's one issue in drawing lines
![image](https://github.com/user-attachments/assets/628da057-3c99-4776-b3bd-8a7a09edefbc)

As you can see in this image, Lines are drawn slightly higher. Even for a single line, it is a bit higher. Any idea?

it gives below warning in console
 ShapeConnectionPin constructor not between 0 and 1.

**Reply by kumilingus** [reply]
Not sure. You can try to debug why `x / width` is not between `0..1` [here](https://github.com/clientIO/joint/blob/master/examples/libavoid/src/avoid-router.js#L199-L200).

**Reply by kumilingus** [reply]
It reads the position of the ports from the model. There is a restriction that the port must be on the boundary of the element for the router to work.
Either redefine your ports or adjust the `ShapeConnectionPin` accordingly.

**Reply by kumilingus** [reply]
You could try increasing the [`shapeBufferDistance`](https://github.com/clientIO/joint/blob/master/examples/libavoid/src/app.js#L228) or rewriting the router to return a bounding box instead of `element.getBBox()`, which returns a rectangle that includes the ports.
